// nojs-react/dist/lib/upload.js
'use strict';

/*
* 文件上传
* nolure@vip.qq.com
* 2013-6-4
*/
var $ = require('jquery'),
    agent = navigator.userAgent.toLowerCase(),
    browser = {
	'0': window.FileReader, //chrome/firefox 使用html5本地预览
	'1': window.ActiveXObject, //ie
	'3': /msie [67]/.test(agent), //ie67		
	'4': window.FormData && window.XMLHttpRequest //FormData
};
browser['2'] = !browser['0'] && !browser['1']; //不支持本地预览


function upload(button, options) {
	button = $(typeof button == 'string' ? '#' + button : button);
	if (!button || !button.length) {
		return;
	}
	var opt = this.options = getOptions(options);

	opt.fileSize = (opt.fileSize || 2) * 1024 * 1024; //默认2M
	opt.multi = opt.limit > 1; //多选模式
	opt.dataField = $.extend(opt.dataField, {
		file: 'file' //上传成功后返回数据中文件url所对应的key
	});

	this.button = button;
	this.parent = this.button.parent();
	this.buttonHTML = button[0].outerHTML;

	this.reset();

	this.tip = typeof opt.tip == 'function' ? opt.tip : function (c) {
		alert(c);
	};
	if (opt.dragUp && opt.dragUp.length) {
		//拖拽上传
		upload.dragUp(this);
	}
	this.init();

	var pasteUpload = opt.pasteUpload;
	if (pasteUpload && pasteUpload.element) {
		upload.pasteUpload(pasteUpload.element, this);
	}
}

var _config = {

	'uploader': null, //文件上传地址
	'getProgressUrl': null, //用于获取文件上传进度的ajax地址
	'name': null, //input文件的name值

	'onSelect': null, //选择文件时
	'onSuccess': null, //上传成功
	'onError': null, //上传失败

	'limit': 10, //文件个数限制
	'fileSize': 2, //文件大小限制，单位MB
	'fileType': '.jpg,.png,.gif,.jpeg', //文件类型后缀
	'showPreview': true, //是否显示缩略图
	'auto': true, //选择文件后是否自动上传   

	'dragUp': null, //拖拽上传，拖拽区域对象

	'showProgress': true, //是否显示进度条

	//上传进度条对象,该对象为公用对象，即每次上传文件时都在该对象内显示进度
	//若需要单独显示每个文件的进度可在onSelect中动态添加dom，内部包含一个class="up_progress"的div即可
	'upProgress': null,
	'fileDomain': '', //上传文件的域名地址
	'dataField': null, //设置一些数据字段
	'data': {} //附加数据

};

function getOptions(options) {
	return $.extend(true, {}, _config, getOptions.fn ? getOptions.fn(options) : options);
}

upload.config = function (options) {
	//全局配置, options可覆盖
	if (typeof options == 'function') {
		getOptions.fn = options;
	} else {
		_config = $.extend(true, _config, options);
	}
};

upload.formatSize = function (size) {
	//格式化文件大小
	if (!size) {
		return null;
	}
	size = size / 1024;
	size = size > 1024 ? (size / 1024).toFixed(2) + 'M' : size.toFixed(2) + 'K';
	return size;
};

upload.dragUp = function (up) {
	var drag = up.options.dragUp;
	if (!window.FileReader) {
		drag.append('您的浏览器不支持拖拽上传');
		return;
	}
	drag = drag[0];
	drag.addEventListener('dragenter', handleDragEnter, false);
	drag.addEventListener('dragover', handleDragOver, false);
	drag.addEventListener('drop', handleFileSelect, false);
	drag.addEventListener('dragleave', handleDragLeave, false);

	// 处理插入拖出效果
	function handleDragEnter(e) {
		e.stopPropagation();
		e.preventDefault();
		$(this).addClass('drag_on');
	}
	function handleDragLeave(e) {
		$(this).removeClass('drag_on');
	}
	// 处理文件拖入事件，防止浏览器默认事件带来的重定向
	function handleDragOver(e) {
		e.stopPropagation();
		e.preventDefault();
		$(this).addClass('drag_on');
	}
	// 处理拖放文件列表
	function handleFileSelect(e) {
		e.stopPropagation();
		e.preventDefault();
		$(this).removeClass('drag_on');

		var files = e.dataTransfer.files,
		    data = { drag: true };
		up.push(files, data);
	}
};

/**
 * 粘贴上传图片
 * 传入文本框对象及上传实例对象
 */
upload.pasteUpload = function (el, up) {

	el && el.addEventListener && el.addEventListener("paste", function (e) {
		for (var i = 0; i < e.clipboardData.items.length; i++) {

			var file = e.clipboardData.items[i];

			if (file.kind == "file" && file.type == "image/png") {
				// get the blob
				var imageFile = file.getAsFile();

				try {
					imageFile.name = +new Date() + "-粘贴上传.png";
				} catch (e) {}

				up.push([imageFile], { from: 'paste' });

				// read the blob as a data URL
				var fileReader = new FileReader();
				fileReader.onloadend = function (e) {
					//console.log(this)                       
				};

				fileReader.readAsDataURL(imageFile);

				// prevent the default paste action
				e.preventDefault();

				// only paste 1 image at a time
				break;
			} else {
				//e.preventDefault()
			}
		}
	});
};

upload.prototype = {
	init: function init(reset) {
		var T = this,
		    data,
		    opt = this.options;

		this.button && reset != false && this.button.remove();
		this.button = $(this.buttonHTML).appendTo(this.parent);
		if (opt.name) {
			this.button[0].name = opt.name;
		}

		if (opt.multi && !/version.+safari/.test(agent)) {
			//safari在多选模式下使用formData上传图片失败
			this.button[0].multiple = true;
		}
		this.button[0].accept = this.options.fileType;
		this.button.change(function () {
			var val = this.value,
			    _file = this.files,
			    type,
			    id;

			if (!val) {
				return;
			}
			//_file = _file.length && Array.prototype.slice.call(_file,0);
			//_file.shift();
			//console.log(val)

			if (T.count >= opt.limit) {
				T.tip('超出数量', 'warn');
				this.value = '';
				return;
			}
			if (T.fileItem[val] && !_file) {
				T.tip('文件已存在', 'warn');
				this.value = '';
				return;
			}
			if (opt.fileType) {
				type = val.substring(val.lastIndexOf('.'), val.length).toLowerCase();
				if (opt.fileType.indexOf(type) < 0) {
					T.tip('文件格式错误', 'warn');
					this.value = '';
					return;
				}
			}
			data = {};
			data.name = val.substring(val.lastIndexOf('\\') + 1, val.length);
			T.push(_file, data);
		});
	},
	push: function push(_file, data) {
		var T = this,
		    n,
		    i,
		    m,
		    size,
		    opt = this.options,
		    val = this.button[0].value;

		data = data || {};
		if (_file) {
			n = _file.length;
			for (i = 0; i < n; i++) {
				if (T.count >= opt.limit) {
					break;
				}
				m = _file[i];
				size = m.size;
				if (size > opt.fileSize) {
					//文件太大
					continue;
				}
				data.size = upload.formatSize(size);
				data.name = m.name;
				//data.type = _file[i].type;
				push(val.substring(0, val.lastIndexOf('\\') + 1) + m.name);
			}
		} else {
			push(val);
		}
		opt.auto && T.startUpload();

		if (browser['4']) {
			this.button.val('');
		}

		function push(val) {
			if (T.fileItem[val]) {
				//文件已存在
				return;
			}
			var id = 'file' + +new Date() + parseInt(Math.random() * 100);

			T.fileItem[val] = data.id = id;

			T.fileItem[id] = $.extend({}, data, {
				file: val,
				button: T.button
			});

			if (_file) {
				T.fileItem[id].files = m;
			}

			T.queue.push(id);

			T.count++;
			T.Events('onSelect', data);

			T.setProgress(0, id, data.size && { loaded: 0, total: '' });

			if (!browser['4'] && !opt.auto) {
				//使用表单上传、非自动上传的情况下，
				T.button.css({
					'position': 'absolute',
					'top': '-999em'
				}).removeAttr('id');
				T.init(false);
			}
			!browser['2'] && T.preview(id);
		}
	},
	/*
  * 所有对外提供的事件集合
  */
	Events: function Events(event, data) {
		var T = this,
		    id,
		    e = T.options[event],
		    args = Array.prototype.slice.call(arguments, 1, arguments.length);

		if (typeof e != 'function') {
			return;
		}
		e = e.apply(this, args);

		if (event == 'onSelect' && e) {
			data = data || {};
			id = data.id;
			T.fileItem[id].wrap = e;

			//给添加的容器绑定移除事件  class='cancel'
			e && e.find('.cancel').click(function (e) {
				// var id = this.id || $(this).data('id')
				T.cancelUpload(id);
				return false;
			});
		}
		return e;
	},
	/*
  * 开始上传
  */
	startUpload: function startUpload() {
		if (this.state || !this.queue.length) {
			return;
		}
		if (typeof this.options.uploader != 'string') {
			this.tip('上传地址错误', 'warn');
			return;
		}

		var T = this,
		    id = this.queue.shift();

		if (!this.fileItem[id]) {
			//该文件已被移除
			return;
		}
		this.id = id; //指向当前正在上传的文件
		this.uploading(id);
	},
	/*
  * 上传处理，使用2种形式
  * 1. 使用html5的FormData api做异步上传
  * 2. 动态添加form对象submit到ifrmae实现无刷新
  */
	uploading: function uploading(id) {
		var T = this,
		    opt = this.options,
		    _file = this.fileItem[id];

		this.state = true;
		_file.state = true; //上传状态

		if (T.Events('beforeUpload', id) == false) {
			return false;
		}

		if (browser['4']) {

			//console.log(xhr.withCredentials)

			var uploadFile = function uploadFile() {
				xhr.upload.addEventListener("progress", uploadProgress, false);
				xhr.addEventListener("load", uploadComplete, false);
				xhr.addEventListener("error", uploadFailed, false);
				xhr.addEventListener("abort", uploadCanceled, false);

				xhr.open("POST", opt.uploader, true);
				//xhr.withCredentials = true;
				xhr.setRequestHeader("X-Requested-With", "XMLHttpRequest");
				var headers = opt.headers || {};
				for (var i in headers) {
					xhr.setRequestHeader(i, headers[i]);
				}

				var fd = new FormData();
				fd.append(T.button[0].name, _file.files);
				fd.append('uploadID', id);
				for (var i in opt.data) {
					fd.append(i, opt.data[i]);
				}
				xhr.send(fd);
			};

			var uploadProgress = function uploadProgress(evt) {
				var percentComplete,
				    p = '';
				if (evt.lengthComputable) {
					percentComplete = Math.round(evt.loaded * 100 / evt.total);
					p = percentComplete.toString();
					T.setProgress(p, null, { loaded: evt.loaded, total: evt.total });
				}
			};

			var uploadComplete = function uploadComplete(evt) {
				//T.setProgress( 100, null );
				T.state = false;
				_file.state = null;
				T.xhr = null;
				if (!evt.target.response) {
					T.Events('onError');
					T.startUpload();
					return;
				}
				var data = evt.target.response ? eval('(' + evt.target.response + ')') : {};
				browser['2'] && T.preview(data); //这里主要针对safari 不能本地预览  
				T.Events('onSuccess', data, id);
				T.startUpload();
			};

			var uploadFailed = function uploadFailed(evt) {
				var data = evt.target.response ? eval('(' + evt.target.response + ')') : {};
				T.Events('onError', data);
			};

			var uploadCanceled = function uploadCanceled(evt) {
				//已中断
			};

			var xhr = this.xhr = new XMLHttpRequest();

			uploadFile();
		} else {
			var name = 'uploadiframe' + +new Date(),
			    callback = opt.jsoncallback || {
				key: 'jsoncallback',
				value: 'upload' + +new Date()
			},
			    action = opt.uploader + (opt.uploader.indexOf('?') < 0 ? '?' : '&'),

			//formData = '<input type="hidden" name="uploadID" value="'+id+'" />',
			formData = '',
			    host = opt.host || location.host;

			opt.data[callback.key] = callback.value;
			for (var i in opt.data) {
				formData += '<input type="hidden" name="' + i + '" value="' + opt.data[i] + '" />';
			}

			this.form = $('<form target="' + name + '" action="' + action + '" method="post" enctype="multipart/form-data" style="position:absolute;left:-999em">' + formData + '</form>').appendTo(document.body);
			this.iframe = $('<iframe name="' + name + '" style="display:none"></iframe>').appendTo(document.body);
			//<script>document.domain="'+host+'"</script>
			//this.iframe[0].contentWindow.document.domain = host;
			_file.button.appendTo(this.form);

			window[callback.value] = function (data) {
				T.init();
				T.form.remove();
				T.iframe.remove();
				T.state = false;
				_file.state = null;
				if (data.status == 1 || !data.error) {
					browser['2'] && T.preview(data);
					T.setProgress(100, T.id);
					T.Events('onSuccess', data, T.id);
				} else {
					T.Events('onError', data, T.id);
				}
				T.startUpload();
				window[callback] = null;
			};
			document.domain = host;
			T.form.submit();
			T.getProgress();
		}
		//!this.queue.length && this.init();
	},
	/*
  * 更新进度条
  */
	setProgress: function setProgress(p, id, size) {
		if (!this.options.showProgress) {
			return;
		}
		var file = this.fileItem[id || this.id],
		    queue = this.options.upProgress || (file && file.wrap ? file.wrap.find('.up_progress') : null);

		queue = typeof queue == 'function' ? queue(p) : queue;
		p += '%';
		//size = size ? '('+upload.formatSize(size.loaded)+'/'+upload.formatSize(size.total)+')' : '';
		queue && queue.html('<div class="progress"><div class="p" style="width:' + p + '"></div><div class="n">' + p + '</div></div>');
	},
	/*
  * 设置本地预览
  * browser['0']：通过html5 api 实现本地预览
  * browser['1']：使用ie滤镜
  * browser['2']：其他浏览器在上传完成返回图片
  */
	preview: function preview(id) {
		if (!this.options.showPreview) {
			return;
		}

		var T = this,
		    _file = T.fileItem[typeof id == 'string' ? id : this.id],
		    file = this.button[0],
		    wrap = _file.wrap;

		if (!wrap || !wrap.length) {
			return;
		}
		if (!/\.(jpg|jpeg|gif|png)$/.test(_file.file)) {
			call('');
			return;
		}
		if (browser['0']) {
			if (_file && _file.files) {
				var reader = new FileReader();
				reader.onload = function (evt) {
					call(evt.target.result);
				};
				reader.readAsDataURL(_file.files);
				//console.log(_file.files.lastModifiedDate.getTime());
				//reader.readAsText(_file.files,'UTF-8')
			}
		} else if (browser['1']) {
			if (!this.options.auto) {
				file = _file.button;
			}
			file.select();
			file.blur();
			_file = document.selection.createRange().text;

			wrap.find('img').replaceWith('<div class="view" style="filter:progid:DXImageTransform.Microsoft.AlphaImageLoader(sizingMethod=\'image\',src=\'' + _file + '\')"></div>');

			//for ie 使用滤镜预览的情况下，等比缩放至合适的尺寸
			setTimeout(function () {
				var img = wrap.find('div.view'),
				    p = img.parent(),
				    W = p.width(),
				    H = p.height(),
				    K = W / H,
				    w = img.width(),
				    h = img.height(),
				    k = w / h;

				if (k > K) {
					w = W;
					h = w / k;
				} else {
					h = H;
					w = k * h;
				}
				img.css({
					'width': w,
					'height': h,
					'filter': 'progid:DXImageTransform.Microsoft.AlphaImageLoader(sizingMethod=\'scale\',src=\'' + _file + '\')"></div>'
				});
				if (browser['3']) {
					//ie67
					p.css('position', 'relative');
					img.css({
						'position': 'absolute',
						'top': '50%',
						'left': '50%',
						'margin-left': -w / 2,
						'margin-top': -h / 2
					});
				}
			}, 100);
		} else {
			//这里针对的主要也是safari 此时id为返回的数据对象
			call(this.options.fileDomain + id[this.options.dataField.file]);
		}
		function call(src) {
			wrap.find('img').attr('src', src);
		}
	},
	//获取上传进度
	getProgress: function getProgress() {
		var url = this.options.getProgressUrl;
		if (!url) {
			return;
		}
		$.ajax({
			url: url,
			data: { 'ajax': 1, 'uploadID': this.id },
			type: 'get',
			dataType: 'text',
			cache: false,
			success: function success(data) {
				//console.log(data);
			}
		});
	},
	//取消上传
	cancelUpload: function cancelUpload(id, reset) {
		if (!id || !this.fileItem[id]) {
			return;
		}
		var T = this,
		    file = this.fileItem[id];

		for (var i = 0; i < this.queue.length; i++) {
			if (this.queue[i] == id) {
				this.queue.splice(i, 1);
				break;
			}
		}
		delete this.fileItem[file.file];
		delete this.fileItem[id];

		this.count--;

		file.wrap && file.wrap.remove();
		file.state && this.xhr && this.xhr.abort(); //终止正在上传的文件

		if (reset) {
			return;
		}
		this.state && this.startUpload(); //继续处理队列

		T.Events('onCancel', file);
	},
	//重置为初始状态
	reset: function reset() {
		// for( var i in this.fileItem ){
		// 	var item = this.fileItem[i]
		// 	item.wrap && item.wrap.remove()
		// 	// this.cancelUpload(i, true)
		// }
		this.fileItem = {}; //保存已上传文件
		this.queue = []; //等待上传队列
		this.state = null; //是否正在上传
		this.count = 0; //已上传文件数
		this.Events('onReset');
	}
};

module.exports = upload;
/*
* bug记录：
* 【已解决】1. auto:false时，移除文件后没有将input值清空,否则无法继续选择刚移除的文件(只针对支持多选的情况)
*/
